src/MQTTClient.php

```diff
@@ -143,29 +155,114 @@ public function connect(
      */
     protected function establishSocketConnection(): void
     {
-        if ($this->hasCertificateAuthorityFile()) {
-            $this->logger->info(sprintf('Using certificate authority file [%s] to verify peer name.', $this->caFile));
-
-            $socketContext = stream_context_create([
-                'ssl' => [
-                'verify_peer_name' => true,
-                    'cafile' => $this->getCertificateAuthorityFile(),
-                ],
-            ]);
-            $connectionString = 'tls://' . $this->getHost() . ':' . $this->getPort();
-            $this->socket     = stream_socket_client($connectionString, $errorCode, $errorMessage, 60, STREAM_CLIENT_CONNECT, $socketContext);
-        } else {
-            $connectionString = 'tcp://' . $this->getHost() . ':' . $this->getPort();
-            $this->socket     = stream_socket_client($connectionString, $errorCode, $errorMessage, 60, STREAM_CLIENT_CONNECT);
+        $useTls = ($this->settings->shouldUseTls() || $this->hasCertificateAuthorityFile());
+        $contextOptions = [];
+
+        if ($useTls) {
+            $contextOptions['ssl'] = [];
+
+            if ($this->settings->shouldTlsVerifyPeer() || $this->hasCertificateAuthorityFile()) {
+                // As it does not make any sense to specify a certificate file without
+                // enabling peer verification, we automatically enable it in that case
+                $contextOptions['ssl']['verify_peer'] = true;
+            }
+            else {
+                // You get a warning output for free, on every connection
+                $this->logger->warning('Enabled TLS encryption without peer verification - DANGEROUS SECURITY ISSUE');
+                $contextOptions['ssl']['verify_peer'] = false;
+            }
+
+            $contextOptions['ssl']['verify_peer_name'] = $this->settings->shouldTlsVerifyPeerName();
+
+            if ($this->hasCertificateAuthorityFile()) {
+                $this->logger->info(sprintf('Using certificate authority file [%s] to verify peer name.', $this->caFile));
```

Comment:
I guess if we touch this, the log statement should also use the CA file getter as used below.

Would probably also be a good opportunity to change this log statement to use the log context in favour of `sprintf`.

Reply:
Sure about the getter, but as much as I am fond of the PSR context, i don't think having a mix of things would be a good idea. Do we go for the full change? Is it worth it to redo the work in 0.2?

Reply:
Hm, better not. It also affects the `SimpleLogger` in the example repository then. It's fixed in the other branch already anyway.